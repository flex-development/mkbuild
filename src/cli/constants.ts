/**
 * @file CLI - Constants
 * @module mkbuild/cli/constants
 */

import pkg from '#pkg' assert { type: 'json' }

/**
 * Array containing choices for boolean flags.
 *
 * @internal
 *
 * @const {string[]} CHOICES_BOOLEAN
 */
export const CHOICES_BOOLEAN: string[] = ['false', 'true', '0', '1']

/**
 * CLI application name.
 *
 * @internal
 *
 * @const {string} CLI_NAME
 */
export const CLI_NAME: string = pkg.name.replace(/.*\//, '')
