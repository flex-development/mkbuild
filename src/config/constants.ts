/**
 * @file Config - Constants
 * @module mkbuild/config/constants
 */

/**
 * Default ignore patterns.
 *
 * @see https://github.com/mrmlnc/fast-glob#ignore
 *
 * @const {string[]} IGNORE_PATTERNS
 */
export const IGNORE_PATTERNS: string[] = [
  '**/.DS_Store',
  '**/.npmignore',
  '**/.yarnignore',
  '**/**.mdx',
  '**/**.stories.{j,t}sx',
  '**/{__mocks__,__snapshots__,__tests__}/**'
]

/**
 * Module extensions checked when attempting to resolve path aliases and module
 * specifiers.
 *
 * @const {string[]} MODULE_EXTENSIONS
 */
export const MODULE_EXTENSIONS: string[] = [
  '.cjs',
  '.cts',
  '.js',
  '.json',
  '.jsx',
  '.mdx',
  '.mjs',
  '.mts',
  '.ts',
  '.tsx'
]

/**
 * `require` and `require.resolve` statement regex.
 *
 * @todo prevent matching non-statements (e.g. comments, arbitrary strings)
 *
 * @const {RegExp} REQUIRE_REGEX
 */
export const REQUIRE_REGEX: RegExp =
  /(?:require\(|require\.resolve\()["']([^"']*)["']\)/g
