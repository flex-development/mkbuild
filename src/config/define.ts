/**
 * @file Config - defineBuildConfig
 * @module mkbuild/config/defineBuildConfig
 */

import type { Config } from '#src/interfaces'

/**
 * Helper for defining a build configuration object.
 *
 * @param {Config} [config={}] - Build configuration options
 * @return {Config} `config`
 */
const defineBuildConfig = (config: Config = {}): Config => config

export default defineBuildConfig
