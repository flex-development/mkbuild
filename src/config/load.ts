/**
 * @file Config - loadBuildConfig
 * @module mkbuild/config/loadBuildConfig
 */

import type { Config } from '#src/interfaces'
import { cosmiconfig } from 'cosmiconfig'
import es from './loader-es'

/**
 * Searches for a build configuration file.
 *
 * If found, a config object will be returned. The object will contain options
 * from the config file. If not found, an empty object will be returned.
 *
 * @see https://github.com/davidtheclark/cosmiconfig#explorersearch
 *
 * @async
 *
 * @param {string} [location] - Directory to search
 * @return {Config} Build configuration options
 */
const loadBuildConfig = async (location: string): Promise<Config> => {
  /**
   * Module name.
   *
   * @const {string} name
   */
  const name: string = 'build'

  // get search function
  const { search } = cosmiconfig(name, {
    loaders: { '.cts': es, '.js': es, '.mjs': es, '.mts': es, '.ts': es },
    searchPlaces: [
      '#.config.cjs',
      '#.config.cts',
      '#.config.js',
      '#.config.json',
      '#.config.mjs',
      '#.config.mts',
      '#.config.ts'
    ].map(place => place.replace('#', name)),
    stopDir: location
  })

  return ((await search(location))?.config as Config | null) ?? {}
}

export default loadBuildConfig
