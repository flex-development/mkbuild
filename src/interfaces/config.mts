/**
 * @file Interfaces - Config
 * @module mkbuild/interfaces/Config
 */

import type { FileSystem, Task } from '@flex-development/mkbuild'

/**
 * Build configuration.
 *
 * @see {@linkcode Task}
 *
 * @extends {Task}
 */
interface Config extends Task {
  /**
   * File system API.
   *
   * @see {@linkcode FileSystem}
   */
  fs?: Partial<FileSystem> | null | undefined

  /**
   * Build tasks.
   *
   * Each build task will be merged with the base task. Top-level lists (arrays
   * and sets) are concatted and deduped. Plain objects are deep merged.
   *
   * > ðŸ‘‰ **Note**: If empty, `null`, or `undefined`, a single build task will
   * > be inferred from the base task.
   *
   * @see {@linkcode Task}
   */
  tasks?: Task[] | null | undefined
}

export type { Config as default }
