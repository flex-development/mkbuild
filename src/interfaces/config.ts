/**
 * @file Interfaces - Config
 * @module mkbuild/interfaces/Config
 */

import type { FileSystemAdapter } from '#src/types'
import type * as esbuild from 'esbuild'
import type Options from './options'
import type Task from './task'

/**
 * Build configuration options.
 *
 * @extends {Options}
 */
interface Config extends Options {
  /**
   * Load build config from config file.
   *
   * @default true
   */
  configfile?: boolean

  /**
   * Build entries.
   *
   * **Note**: If empty or `undefined`, a single build entry will be inferred
   * from the remaining set of configuration options.
   *
   * @default []
   */
  entries?: Partial<Omit<Task, 'write'>>[]

  /**
   * Custom implementations of file system methods.
   *
   * @see {@linkcode FileSystemAdapter}
   * @see https://nodejs.org/api/fs.html
   *
   * @default fsa
   */
  fs?: FileSystemAdapter

  /**
   * Serve files.
   *
   * @see {@linkcode esbuild.ServeOptions}
   * @see https://esbuild.github.io/api/#serve
   *
   * @default false
   */
  serve?: esbuild.ServeOptions | boolean

  /**
   * Watch files.
   *
   * @default false
   */
  watch?: boolean

  /**
   * Write build results.
   *
   * @default false
   */
  write?: boolean
}

export type { Config as default }
