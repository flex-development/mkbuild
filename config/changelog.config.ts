#!/usr/bin/env node

/**
 * @file Configuration - Changelog
 * @module config/changelog
 * @see https://github.com/conventional-changelog/conventional-changelog/tree/master/packages/conventional-changelog-conventionalcommits
 * @see https://github.com/conventional-changelog/conventional-changelog/tree/master/packages/conventional-changelog-core
 * @see https://github.com/conventional-changelog/conventional-changelog/tree/master/packages/conventional-changelog-writer
 * @see https://github.com/conventional-changelog/conventional-changelog/tree/master/packages/conventional-commits-parser
 * @see https://github.com/conventional-changelog/conventional-changelog/tree/master/packages/git-raw-commits
 */

import { HelpService, UtilityService } from '#src/cli/providers'
import {
  Type,
  parserPreset,
  type Commit
} from '@flex-development/commitlint-config'
import pathe from '@flex-development/pathe'
import { CompareResult, isNIL } from '@flex-development/tutils'
import { Inject, Module } from '@nestjs/common'
import addStream from 'add-stream'
import { Command, CommanderError } from 'commander'
import consola from 'consola'
import conventionalChangelog from 'conventional-changelog'
import type { Options } from 'conventional-changelog-core'
import type {
  CommitGroup,
  GeneratedContext
} from 'conventional-changelog-writer'
import dateformat from 'dateformat'
import {
  CommandFactory,
  CommandRunner,
  Option,
  RootCommand
} from 'nest-commander'
import {
  createReadStream,
  createWriteStream,
  readFileSync,
  type ReadStream,
  type WriteStream
} from 'node:fs'
import type { Readable } from 'node:stream'
import semver from 'semver'
import tempfile from 'tempfile'
import pkg from '../package.json' assert { type: 'json' }

/**
 * Parsed commit with additional fields.
 *
 * @extends {Commit}
 */
interface CommitEnhanced extends Commit {
  raw: Commit
  version?: string
}

/**
 * CLI options.
 */
interface Flags {
  /**
   * Enable verbose output.
   *
   * @default false
   */
  debug?: boolean

  /**
   * Print help text.
   */
  help?: boolean

  /**
   * Read `CHANGELOG` from this file.
   */
  infile?: string

  /**
   * Write content to this file.
   *
   * **Note**: Requires {@linkcode write} to be `true`.
   */
  outfile?: string

  /**
   * Output unreleased changelog.
   *
   * To the set the current version, pass a string value.
   *
   * @example
   *  '1.0.0-alpha.7'
   * @example
   *  true
   *
   * @default true
   */
  outputUnreleased?: boolean | string

  /**
   * Number of releases to be generated.
   *
   * If `0`, the changelog will be regenerated and the output file will be
   * overwritten.
   *
   * @default 1
   */
  releaseCount?: number

  /**
   * Output content to {@linkcode infile}.
   *
   * @default false
   */
  samefile?: boolean

  /**
   * Write content to file instead of {@linkcode process.stdout}.
   *
   * @default false
   */
  write?: boolean
}

/**
 * `changelog` command model.
 *
 * @class
 * @extends {CommandRunner}
 */
@RootCommand({ description: 'Changelog CLI' })
class ChangelogCommand extends CommandRunner {
  /**
   * Creates a new `changelog` command instance.
   *
   * @param {HelpService} help - Help service instance
   * @param {UtilityService} util - CLI utilities service instance
   */
  constructor(
    @Inject(HelpService) protected readonly help: HelpService,
    @Inject(UtilityService) protected readonly util: UtilityService
  ) {
    super()
  }

  /**
   * Parses the `--debug` flag.
   *
   * @see {@linkcode Flags.debug}
   *
   * @protected
   *
   * @return {true} Option value
   */
  @Option({
    defaultValue: false,
    description: 'Enable verbose output',
    flags: '-d, --debug',
    name: 'debug'
  })
  protected parseDebug(): true {
    return true
  }

  /**
   * Parses the `--help` flag.
   *
   * @see {@linkcode Flags.help}
   *
   * @protected
   *
   * @return {true} Option value
   */
  @Option({
    description: 'Display this message',
    flags: '-h, --help',
    name: 'help'
  })
  protected parseHelp(): true {
    return true
  }

  /**
   * Parses the `--infile` flag.
   *
   * @see {@linkcode Flags.infile}
   *
   * @protected
   *
   * @param {string} val - Value to parse
   * @return {string} Parsed option value
   */
  @Option({
    description: 'Read CHANGELOG from this file',
    flags: '-i, --infile <path>',
    name: 'infile'
  })
  protected parseInfile(val: string): string {
    return val
  }

  /**
   * Parses the `--outfile` flag.
   *
   * @see {@linkcode Flags.outfile}
   *
   * @protected
   *
   * @param {string} val - Value to parse
   * @return {string} Parsed option value
   */
  @Option({
    description: 'Write content to this file (requires --write)',
    flags: '-o, --outfile <path>',
    name: 'outfile'
  })
  protected parseOutfile(val: string): string {
    return val
  }

  /**
   * Parses the `--output-unreleased` flag.
   *
   * @see {@linkcode Flags.outputUnreleased}
   *
   * @protected
   *
   * @param {string} val - Value to parse
   * @return {boolean | string} Parsed option value
   */
  @Option({
    defaultValue: true,
    description: 'Output unreleased changelog',
    flags: '-u, --output-unreleased [option]',
    name: 'outputUnreleased'
  })
  protected parseOutputUnreleased(val: string): boolean | string {
    try {
      return this.util.parseBoolean(val)
    } catch {
      return val
    }
  }

  /**
   * Parses the `--release-count` flag.
   *
   * @see {@linkcode Flags.releaseCount}
   *
   * @protected
   *
   * @param {string} val - Value to parse
   * @return {number} Parsed option value
   */
  @Option({
    defaultValue: 1,
    description: 'Number of releases to be generated',
    flags: '-r, --release-count <count>',
    name: 'releaseCount'
  })
  protected parseReleaseCount(val: string): number {
    return this.util.parseInt(val)
  }

  /**
   * Parses the `--samefile` flag.
   *
   * @see {@linkcode Flags.samefile}
   *
   * @protected
   *
   * @return {true} Option value
   */
  @Option({
    defaultValue: false,
    description: 'Output content to infile',
    flags: '-s, --samefile',
    name: 'samefile'
  })
  protected parseSamefile(): true {
    return true
  }

  /**
   * Parses the `--write` flag.
   *
   * @see {@linkcode Flags.write}
   *
   * @protected
   *
   * @return {true} Option value
   */
  @Option({
    defaultValue: false,
    description: 'Write content to file',
    flags: '-w, --write',
    name: 'write'
  })
  protected parseWrite(): true {
    return true
  }

  /**
   * Command runner.
   *
   * @public
   * @async
   *
   * @param {string[]} _ - Command arguments
   * @param {Flags} [flags={}] - Command options
   * @return {Promise<void>} Nothing when complete
   */
  public async run(_: string[], flags: Flags = {}): Promise<void> {
    const { debug, help, outputUnreleased, releaseCount, write } = flags
    let { infile, outfile, samefile } = flags

    // print help text and exit
    if (help) return void consola.log(this.help.formatHelp(this.command))

    /**
     * Regex used to extract a release version from a string containing
     * Git tags.
     *
     * @const {RegExp} vgx
     */
    const vgx: RegExp = pkg.tagPrefix
      ? new RegExp(`tag:\\s*[=]?${pkg.tagPrefix}(.+?)[,)]`, 'gi')
      : /tag:\s*[=v]?(.+?)[),]/gi

    /**
     * Changelog content stream.
     *
     * @const {Readable} changelog
     */
    // @ts-expect-error type definitions are incorrect
    const changelog: Readable = conventionalChangelog<Commit>(
      {
        append: false,
        debug: debug ? consola.log.bind(consola) : undefined,
        outputUnreleased:
          typeof outputUnreleased === 'boolean'
            ? outputUnreleased
            : typeof outputUnreleased === 'string'
            ? !!outputUnreleased.trim()
            : false,
        pkg: { path: pathe.resolve('package.json') },
        preset: {
          header: '',
          name: parserPreset.name,
          releaseCommitMessageFormat: 'release: {{currentTag}}',
          types: [
            { section: ':package: Build', type: Type.BUILD },
            { section: ':house_with_garden: Housekeeping', type: Type.CHORE },
            { section: ':robot: Continuous Integration', type: Type.CI },
            { section: ':pencil: Documentation', type: Type.DOCS },
            { section: ':sparkles: Features', type: Type.FEAT },
            { section: ':bug: Fixes', type: Type.FIX },
            { section: ':fire: Performance Improvements', type: Type.PERF },
            { section: ':mechanical_arm: Refactors', type: Type.REFACTOR },
            { section: ':wastebasket: Reverts', type: Type.REVERT },
            { hidden: true, type: Type.STYLE },
            { section: ':white_check_mark: Testing', type: Type.TEST },
            { hidden: true, type: Type.WIP }
          ]
        },
        releaseCount,
        skipUnstable: false,
        tagPrefix: pkg.tagPrefix,
        /**
         * Raw commit transformer.
         *
         * @see https://github.com/conventional-changelog/conventional-changelog/tree/master/packages/conventional-changelog-core#transform-1
         * @see https://github.com/conventional-changelog/conventional-changelog/issues/415
         *
         * @param {Commit} commit - Commit object
         * @param {Options.Transform.Callback} apply - Commit handler
         * @return {void} Nothing when complete
         */
        transform(commit: Commit, apply: Options.Transform.Callback): void {
          return void apply(null, {
            ...commit,
            committerDate: dateformat(commit.committerDate, 'yyyy-mm-dd', true),
            mentions: commit.mentions.filter(m => m !== 'flexdevelopment'),
            // @ts-expect-error ts(2322)
            raw: commit,
            references: commit.references.filter(ref => ref.action !== null),
            version: commit.gitTags ? vgx.exec(commit.gitTags)?.[1] : undefined
          })
        },
        warn: parserPreset.parserOpts.warn
      },
      {},
      {
        debug: debug ? parserPreset.parserOpts.warn : undefined,
        format:
          '%B%n-hash-%n%H%n-shortHash-%n%h%n-gitTags-%n%d%n-committerDate-%n%ci%n'
      },
      parserPreset.parserOpts,
      {
        /**
         * Sorts commit groups in descending order by group title.
         *
         * GitHub emojis in titles will be ignored.
         *
         * @see https://github.com/conventional-changelog/conventional-changelog/tree/master/packages/conventional-changelog-writer#commitgroupssort
         *
         * @param {CommitGroup} a - Commit group object
         * @param {CommitGroup} b - Commit group object to compare to `a`
         * @return {number} Compare result
         */
        commitGroupsSort(a: CommitGroup, b: CommitGroup): number {
          if (a.title === false) return CompareResult.GREATER_THAN
          if (b.title === false) return CompareResult.LESS_THAN

          /**
           * Regex used to extract commit group titles without GitHub emojis.
           *
           * @const {RegExp} tgx - Regex used to extract commit group title
           */
          const tgx: RegExp = /([A-Z])\w+/

          return tgx.exec(a.title)![0]!.localeCompare(tgx.exec(b.title)![0]!)
        },
        /**
         * Sorts commits in descending order by commit header and date.
         *
         * @see https://github.com/conventional-changelog/conventional-changelog/tree/master/packages/conventional-changelog-writer#commitssort
         *
         * @param {Commit} a - Commit object
         * @param {Commit} b - Commit object to compare to `b`
         * @return {number} Compare result
         */
        commitsSort(a: Commit, b: Commit): number {
          /**
           * Compare result for {@linkcode b.committerDate} and
           * {@linkcode a.committerDate}.
           *
           * @const {number} by_date
           */
          const by_date: number =
            new Date(b.committerDate).getTime() -
            new Date(a.committerDate).getTime()

          return a.header && b.header
            ? a.header.localeCompare(b.header) || by_date
            : by_date
        },
        /**
         * Modifies `context` before the changelog is generated.
         *
         * This includes:
         *
         * - Setting the current and previous release tags
         * - Setting the release date
         * - Determining patch release state
         * - Determining if compare links should be generated
         *
         * @see https://github.com/conventional-changelog/conventional-changelog/tree/master/packages/conventional-changelog-core#finalizecontext
         * @see https://github.com/conventional-changelog/conventional-changelog/tree/master/packages/conventional-changelog-writer#finalizecontext
         *
         * @param {GeneratedContext} context - Generated changelog context
         * @param {Options} options - `conventional-changelog-core` options
         * @param {CommitEnhanced[]} commits - Commits for release
         * @param {CommitEnhanced?} key - Release commit
         * @return {GeneratedContext} Final changelog context
         */
        finalizeContext(
          context: GeneratedContext,
          options: Options,
          commits: CommitEnhanced[],
          key?: CommitEnhanced
        ): GeneratedContext {
          const { gitSemverTags = [], isPatch, linkCompare, version } = context
          let { currentTag, previousTag } = context

          /**
           * First commit in release.
           *
           * @const {CommitEnhanced?} first_commit
           */
          const first_commit: CommitEnhanced | undefined = commits.at(0)

          /**
           * Last commit in release.
           *
           * @const {CommitEnhanced?} last_commit
           */
          const last_commit: CommitEnhanced | undefined = commits.at(-1)

          // set current and previous tags
          if (key && (!currentTag || !previousTag)) {
            currentTag = key.version ?? undefined

            // try setting previous tag based on current tag
            if (gitSemverTags.includes(currentTag ?? '')) {
              const { version = '' } = key
              previousTag = gitSemverTags[gitSemverTags.indexOf(version) + 1]
              if (!previousTag) previousTag = last_commit?.hash ?? undefined
            }
          } else {
            currentTag = /^unreleased$/i.test(version ?? '')
              ? currentTag ??
                (typeof outputUnreleased === 'string' && outputUnreleased
                  ? outputUnreleased
                  : first_commit?.hash ?? undefined)
              : !currentTag && version
              ? pkg.tagPrefix + version
              : currentTag ?? version
            previousTag = previousTag ?? gitSemverTags[0]
          }

          // set release date
          context.date =
            key?.committerDate ??
            dateformat(new Date().toLocaleDateString(), 'yyyy-mm-dd', true)

          // determine patch release state
          if (version && semver.valid(version)) {
            context.isPatch = isPatch ?? semver.patch(version) !== 0
          }

          // @ts-expect-error ts(2322)
          return {
            ...context,
            currentTag,
            linkCompare: isNIL(linkCompare) && !!currentTag && !!previousTag,
            previousTag,
            repoUrl: pkg.repository.slice(0, -4)
          }
        },
        headerPartial: readFileSync(
          'config/templates/changelog/header.hbs',
          'utf8'
        ),
        ignoreReverted: false
      }
    ).on('error', err => consola.error(err.stack))

    // override samefile if infile and outfile are the same file
    if (infile && infile === outfile) samefile = true

    // reset outfile if changelog should be output to infile
    if (samefile) outfile = infile = infile ?? 'CHANGELOG.md'

    /**
     * Creates a file writer.
     *
     * If writing to file is disabled, {@linkcode process.stdout} will be used
     * to write content to the terminal.
     *
     * Otherwise, {@linkcode createWriteStream} will be used to create a stream
     * for {@linkcode outfile}.
     *
     * @see {@linkcode NodeJS.WriteStream}
     * @see {@linkcode WriteStream}
     * @see {@linkcode createWriteStream}
     *
     * @return {WriteStream | (NodeJS.WriteStream & { fd: 1 })} File writer
     */
    const writer = (): WriteStream | (NodeJS.WriteStream & { fd: 1 }) => {
      return write && outfile ? createWriteStream(outfile) : process.stdout
    }

    // write changelog to infile, outfile, or stdout
    switch (true) {
      case infile && releaseCount !== 0:
        /**
         * Changelog file stream.
         *
         * @const {ReadStream} rs
         */
        const rs: ReadStream = createReadStream(infile!).on('error', () => {
          if (debug) consola.error('infile does not exist.')
          if (samefile) changelog.pipe(writer())
        })

        // write changelog to infile or stdout
        if (samefile) {
          /**
           * Absolute path to random temporary file.
           *
           * @const {string} tmp
           */
          const tmp: string = tempfile()

          changelog
            .pipe(addStream(rs))
            .pipe(createWriteStream(tmp))
            .on('finish', () => createReadStream(tmp).pipe(writer()))
        } else {
          changelog.pipe(addStream(rs)).pipe(writer())
        }

        break
      default:
        changelog.pipe(writer())
        break
    }

    return void 0
  }

  /**
   * Sets {@linkcode command}.
   *
   * @protected
   * @override
   *
   * @param {Command} cmd - CLI command instance
   * @return {this} `this`
   */
  public override setCommand(cmd: Command): this {
    cmd.addHelpCommand(false)
    cmd.allowExcessArguments(false)
    cmd.allowUnknownOption(false)
    cmd.createHelp = () => this.help
    cmd.combineFlagAndOptionalValue(false)
    cmd.enablePositionalOptions()
    cmd.helpOption(false)
    cmd.showHelpAfterError()
    cmd.showSuggestionAfterError()
    this.command = cmd
    return this
  }
}

/**
 * CLI application module.
 *
 * @class
 */
@Module({ providers: [ChangelogCommand, HelpService, UtilityService] })
class AppModule {
  /**
   * Commander error handler.
   *
   * @see {@linkcode Command.exitOverride}
   * @see {@linkcode CommanderError}
   * @see https://github.com/jmcdo29/nest-commander/blob/nest-commander%403.6.1/packages/nest-commander/src/command-runner.service.ts#L49-L51
   *
   * @public
   * @static
   *
   * @param {CommanderError} error - Error to handle
   * @return {void} Nothing when complete
   */
  public static errorHandler(error: CommanderError): void {
    if (error.exitCode) consola.error(error)
    return void process.exit(error.exitCode)
  }

  /**
   * CLI command error handler.
   *
   * @public
   * @static
   *
   * @param {Error} error - Error to handle
   * @return {void} Nothing when complete
   */
  public static serviceErrorHandler(error: Error): void {
    consola.error(error)
    return void process.exit(1)
  }
}

/**
 * CLI application runner.
 *
 * @async
 *
 * @return {Promise<void>} Nothing when complete
 */
async function main(): Promise<void> {
  return void (await CommandFactory.run(AppModule, {
    errorHandler: AppModule.errorHandler,
    logger: ['error', 'warn'],
    serviceErrorHandler: AppModule.serviceErrorHandler
  }))
}

// run application and exit
void (await main())
