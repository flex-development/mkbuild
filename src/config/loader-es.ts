/**
 * @file Config - esLoader
 * @module mkbuild/config/esLoader
 */

import type { Config } from '#src/interfaces'
import tsconfigPaths from '#src/plugins/tsconfig-paths/plugin'
import { build } from 'esbuild'
import { resolveImports, toDataURL } from 'mlly'
import { RESOLVE_EXTENSIONS } from './constants'

/**
 * Loads an [ESM][1] (`*.js`, `*.mjs`) or TypeScript (`*.cts`, `*.mts`, `*.ts`)
 * build config file.
 *
 * [1]: https://nodejs.org/api/esm.html
 *
 * @see https://github.com/davidtheclark/cosmiconfig#loaders
 *
 * @async
 *
 * @param {string} path - Config file path
 * @param {string} content - Config file content
 * @return {Promise<Config>} Build configuration options
 */
const esLoader = async (path: string, content: string): Promise<Config> => {
  const {
    outputFiles: [output]
  } = await build({
    entryPoints: [path],
    format: 'esm',
    loader: {
      '.cts': 'ts',
      '.js': 'js',
      '.mjs': 'js',
      '.mts': 'ts',
      '.ts': 'ts'
    },
    metafile: true,
    outdir: '<stdout>',
    plugins: [tsconfigPaths()],
    write: false
  })

  // resolve imports in content to evaluate module
  content = await resolveImports(output!.text, {
    extensions: RESOLVE_EXTENSIONS,
    url: path
  })

  return ((await import(toDataURL(content))) as { default: Config }).default
}

export default esLoader
