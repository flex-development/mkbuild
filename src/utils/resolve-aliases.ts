/**
 * @file Utilities - resolveAliases
 * @module mkbuild/utils/resolveAliases
 */

import { MODULE_EXTENSIONS } from '#src/config/constants'
import type { Entry } from '#src/interfaces'
import type { Metafile, OutputFile } from 'esbuild'
import type { MatchPath } from 'tsconfig-paths'
import type { ReadJsonSync } from 'tsconfig-paths/lib/filesystem'
import extractStatements from './extract-statements'
import resolveAlias from './resolve-alias'

/**
 * Resolves path aliases in `output` file content.
 *
 * @param {OutputFile} output - Output file object
 * @param {Uint8Array} output.contents - Output file content
 * @param {string} output.path - Absolute path to output file
 * @param {string} output.text - `output.contents` as string
 * @param {Pick<Metafile, 'outputs'>} metadata - Build metadata
 * @param {Entry['format']} format - Output file format
 * @param {MatchPath} matcher - Path matching function
 * @param {string} [cwd=process.cwd()] - Current working directory
 * @param {string[]} [extensions=MODULE_EXTENSIONS] - Extensions to probe for
 * @param {(path: string) => boolean} [fileExists] - File existence checker
 * @param {ReadJsonSync} [readJson] - Reads `JSON` data from a file
 * @return {OutputFile} `output` with path aliases in file content resolved
 */
const resolveAliases = (
  output: OutputFile,
  metadata: Pick<Metafile, 'outputs'>,
  format: Entry['format'],
  matcher: MatchPath,
  cwd: string = process.cwd(),
  extensions: string[] = MODULE_EXTENSIONS,
  fileExists?: (path: string) => boolean,
  readJson?: ReadJsonSync
): OutputFile => {
  /**
   * Relative path to output file.
   *
   * **Note**: Relative to {@link cwd}.
   *
   * @const {string} outfile
   */
  const outfile: string = output.path.replace(cwd + '/', '')

  // do nothing if output does not have metadata
  if (!metadata.outputs[outfile]) return output

  // do nothing if output does not have entry point
  if (!metadata.outputs[outfile]!.entryPoint) return output

  /**
   * {@link output.text} copy.
   *
   * @var {string} text
   */
  let text: string = output.text

  for (const statement of extractStatements(output.text, format)) {
    // do nothing if missing module specifier
    if (!statement.specifier) continue

    /**
     * Possible path match for {@link statement.specifier}.
     *
     * @const {string | undefined} match
     */
    const match: string | undefined = matcher(
      statement.specifier,
      readJson,
      fileExists,
      extensions
    )

    // do nothing if path match was not found
    if (!match) continue

    /**
     * {@link statement.code} before path alias replacement.
     *
     * @const {string} code
     */
    const code: string = statement.code

    // resolve path alias in statement
    resolveAlias(statement, match, metadata.outputs[outfile]!.entryPoint!)

    // replace code in text
    text = text.replace(code, statement.code)
  }

  return { ...output, contents: new Uint8Array(Buffer.from(text)), text }
}

export default resolveAliases
