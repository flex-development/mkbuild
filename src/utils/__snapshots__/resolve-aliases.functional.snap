// Vitest Snapshot v1

exports[`functional:utils/resolveAliases > resolve > should resolve path aliases in output file content 1`] = `
"import { MODULE_EXTENSIONS } from '../config/constants'
import extractStatements from './extract-statements'
import resolveAlias from './resolve-alias'
import type { Format } from 'esbuild'
import type { MatchPath } from 'tsconfig-paths'
import type { ReadJsonSync } from 'tsconfig-paths/lib/filesystem'"
`;
